llama-rs/src/lib.rs

```diff
@@ -1142,4 +1148,95 @@ impl Model {
 
         res
     }
+
+    /// Obtains a serializable snapshot of the current inference status. This
+    /// can be used to cache the state of the model and store them into a file.
+    ///
+    /// # Safety
+    ///
+    /// This function provides raw access to the underlying memory owned by the
+    /// ggml context. While the provided `LlamaModelMemory` object is alive, no
+    /// other methods for this model object should be called.
+    pub unsafe fn get_snapshot(&mut self) -> InferenceSnapshotRef<'_> {
+        use core::slice;
+        let memory_k = unsafe {
+            slice::from_raw_parts(self.memory_k.data() as *mut u8, self.memory_k.nbytes())
+        };
+        let memory_v = unsafe {
+            slice::from_raw_parts(self.memory_v.data() as *mut u8, self.memory_v.nbytes())
+        };
+
+        InferenceSnapshotRef {
+            npast: self.n_past,
+            memory_k,
+            memory_v,
+            last_n_tokens: self.last_n_tokens.clone(),
+            logits: self.last_logits.clone(),
+        }
+    }
+
+    /// Clears the model's state, so inference can be restarted without having
+    /// to load the weights again.
+    pub fn clear_state(&mut self) {
+        self.memory_k.zero_data();
+        self.memory_v.zero_data();
+        self.n_past = 0;
+        self.last_n_tokens = VecDeque::from(vec![0; self.last_n_tokens.len()]);
+    }
+
+    /// Sets the state of the model, from a previously obtained InfrenceSnapshot
+    pub fn set_snapshot(&mut self, snapshot: InferenceSnapshot) -> Result<(), LlamaError> {
+        if self.memory_k.nbytes() != snapshot.memory_k.len()
+            || self.memory_v.nbytes() != snapshot.memory_v.len()
+        {
+            return Err(LlamaError::MemorySizeMismatch {
+                self_size: self.memory_k.nbytes() + self.memory_v.nbytes(),
+                input_size: snapshot.memory_k.len() + snapshot.memory_v.len(),
+            });
+        }
+
+        // SAFETY: We have exclusive access to Self, which means no one else
+        // should be touching the context's memory. We can write to it because
+        // we already checked the size.
+        unsafe {
+            self.memory_k.write_data(&snapshot.memory_k);
+            self.memory_v.write_data(&snapshot.memory_v);
+        }
+
+        self.n_past = snapshot.npast;
+        self.last_n_tokens = snapshot.last_n_tokens;
+        self.last_logits = snapshot.last_logits;
+
+        Ok(())
+    }
+}
+
+impl<'a> InferenceSnapshotRef<'a> {
+    pub fn write_to_disk(&self, path: impl AsRef<Path>) -> Result<(), LlamaError> {
```

Comment:
Suggest splitting this into `write(&self, writer: &mut Write)` and `write_to_disk`, where the latter just calls the former with a `File`